servicetalk-http-api/src/main/java/io/servicetalk/http/api/HttpClientBuilder.java

```diff
@@ -156,11 +159,18 @@
             Function<U, CharSequence> unresolvedAddressToHostFunction);
 
     /**
-     * Disables automatically delaying {@link StreamingHttpRequest}s until the {@link LoadBalancer} is ready.
+     * Updates the automatic retry strategy for the clients generated by this builder. Automatic retries are done by
+     * the clients automatically when allowed by the passed {@link AutoRetryStrategyProvider}. These retries are not a
+     * substitute for user level retries which are designed to infer retry decisions based on request/error information.
+     * Typically such user level retries are done using filters (eg:
```

Comment:
It's good that we're trying to provide more context, but I feel like the description is too vague in what an auto-retry could/should be used for. It's heavily tied to load-balancing from an API perspective already, maybe we should just call it out.  ?

Reply:
I hear what you are saying but there is a balance about prescribing implementation details at this level or staying neutral to have the strategy define the behaviours. I thought this language was somehow walking that line. 

The API requires a `LoadBalancer` as the LB events are not otherwise accessible. However, it is not limited to load balancer as we are retrying transport errors.  

Do you think you can suggest an alternative language? I will be happy to update.

Reply:
We need to address the underlying problem which is the real implementation quirk users shouldn't have to know about. It's not clear to me why we'd want to disable/configure this and LB-ready. More so it's surprising to users that we don't do this behind the scenes as it's not under their control and all other libraries do this.  

Maybe you envisioned this API enabling other features in the future, but I'm not yet convinced this kind of abstraction is helpful for a user. At the very least it should be clear what _kind_ of behaviors to expect or not-expect should people want to implement their own. We traded an implementation detail for vagueness in this PR and neither is ideal. If we're not sure or not comfortable with describing the intended uses, maybe reconsider introducing as public API.  

As far as wording, I'd consider something like:
> The Client consults this strategy in case a request couldn't be made, for ex. due to state of the LoadBalancer or its Connection(s). It may allow the Client to automatically retry requests that failed when it is guaranteed safe to do so according to the strategy. Not all failures caused by an issue with the Connection are safe to retry without side-effects and therefore application-level error handling is recommended.

Reply:
I also have mixed fillings about this API. It's nice that users can control and disable these retries, but do we know the use case when it's necessary? Can we defer this level of control until we know why these retries cause a problem? And until that just do it internally. It looks like for 99.9% of cases users will be happy if ST solves these type of problems automatically.

I'm agreed that we may want to increase level of visibility for these type of exceptions and notify users about what is happening under the hood. It may help them debug server-side or service discovery problems. But for disabling these retries there are other ways: they can always wrap LB interface and this wrapper may wrap specific exceptions to hide the `RetryableException` marker. Or, if `NoAvailableHostException` causes undesired high latency for the first request users can apply timeout filter.

Reply:
I may not have been clear about this in the description but 

__This is not a problem due to ServiceTalk design, it is a problem due to concurrent actions from users and network__

The intentional vaguenss in the builder methods is to defer the details to the supplied strategy. `DefaultAutoRetryStrategyProvider` gives relevant details about the behaviors, eg:

```java
        /**
         * Connection closures (by the peer or locally) and new requests may happen concurrently. This means that it is
         * possible for a {@link LoadBalancer} to select a connection which is already closed (concurrently) but the
         * close signal has not yet been seen the {@link LoadBalancer}. In such cases, requests fail with a
         * {@link RetryableException}. By default, automatic retries always retries these {@link RetryableException}s.
         * This method disables the default behavior.
         *
         * @return {@code this}.
         */
        public Builder disableRetryAllRetryableExceptions() {
```

So I would not say that "we're not sure or not comfortable with describing the intended uses" but just being less opinionated on a generic API with deferring specifics to specific API.

As a reference grpc-java has similar control knobs in form of [waitForReady](https://github.com/grpc/grpc-java/blob/ce2b0174b2be207cb316faf49e923afe09e0f264/api/src/main/java/io/grpc/CallOptions.java#L156-L160) which we already provided before this PR in the form of `disableWaitForLoadBalancer`. When disabled, the clients display a "fail-fast" behavior which may be required in cases to reject traffic before the system is warmed up (indicated through composing multiple client statuses) as opposed to holding requests in memory.
 
As a principle, I think if we make a decision to provide default for a behavior which isn't otherwise stated in the specification, we should provide a way to disable the behavior when reasonably possible. The motivation to extract a separate interface is to consolidated auto-retry behaviors for different type of failures which seems better than providing two top-level methods that are non-intuitively related.

Reply:
After thinking about your comment and discussing offline it makes sense to expose this config. It helps with discoverability of this behavior to control it and gives a way to configure max number of retries that could be useful in envs with slow ServiceDiscovery.